src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -156,6 +201,492 @@ MultiRouterPlanCreate(Query *originalQuery, Query *query,
+ * INSERT ... SELECT queries.
+ */
+static void
+ErrorIfInsertSelectQueryNotSupported(Query *queryTree)
+{
+	RangeTblEntry *subqueryRte = NULL;
+	Query *subquery = NULL;
+
+	/* we only do this check for INSERT ... SELECT queries */
+	AssertArg(InsertSelectQuery(queryTree));
+
+	subqueryRte = lsecond(queryTree->rtable);
+	subquery = subqueryRte->subquery;
+
+	/* we support this feature only for colocated tables */
+	ErrorIfNotAllParticipatingTablesAreColocated(queryTree);
+
+	if (contain_mutable_functions((Node *) queryTree))
+	{
+		ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						errmsg("cannot perform distributed planning for the given "
+							   "modification"),
+						errdetail(
+							"Stable and volatile functions are not allowed in INSERT ... "
+							"SELECT queries")));
+	}
+
+	if (queryTree->cteList != NULL)
+	{
+		ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						errmsg("cannot perform distributed planning for the given "
+							   "modification"),
+						errdetail(
+							"Common table expressions are not allowed in INSERT ... SELECT "
+							"queries")));
+	}
+
+	/* we don't support LIMIT, OFFSET and WINDOW functions */
+	ErrorIfMultiTaskRouterSelectQueryUnsupported(subquery);
+
+	/* ensure that INSERT's partition column comes from SELECT's partition column */
+	ErrorIfInsertPartitionColumnDoesNotMatchSelect(queryTree);
+}
+
+
+/*
+ *  ErrorUnsupportedMultiTaskSelectQuery errors out on queries that we support
+ *  for single task router queries, but, cannot allow for multi task router
+ *  queries. We do these checks recursively to prevent any wrong results.
+ */
+static void
+ErrorIfMultiTaskRouterSelectQueryUnsupported(Query *query)
+{
+	List *queryList = NIL;
+	ListCell *queryCell = NULL;
+
+	ExtractQueryWalker((Node *) query, &queryList);
+	foreach(queryCell, queryList)
+	{
+		Query *subquery = (Query *) lfirst(queryCell);
+
+		Assert(subquery->commandType == CMD_SELECT);
+
+		if (subquery->limitCount != NULL)
+		{
+			ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+							errmsg("cannot perform distributed planning for the given "
+								   "modification"),
+							errdetail(
+								"LIMIT clauses are not allowed in INSERT ... SELECT "
+								"queries")));
+		}
+
+		if (subquery->limitOffset != NULL)
+		{
+			ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+							errmsg("cannot perform distributed planning for the given "
+								   "modification"),
+							errdetail(
+								"OFFSET clauses are not allowed in INSERT ... SELECT "
+								"queries")
+
+							));
+		}
+
+		if (subquery->windowClause != NULL)
+		{
+			ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+							errmsg("cannot perform distributed planning for the given "
+								   "modification"),
+							errdetail(
+								"Window functions are not allowed in INSERT ... SELECT "
+								"queries")));
+		}
+
+		if (subquery->setOperations != NULL)
```

Comment:
I'm still looking into it.

We end up with calling `ConvertRteToSubqueryWithEmptyResult()` while `UpdateRelationNames()` is running. In the end, I got the following queries which :

```SQL

INSERT INTO agg_events 
            (value_4_agg, 
             value_1_agg, 
             user_id) 
(SELECT value_4, 
        value_1, 
        user_id 
 FROM   raw_events_first 
 WHERE  user_id = 8) 
UNION 
(SELECT value_4, 
        value_1, 
        user_id 
 FROM   raw_events_second 
 WHERE  user_id = 8); 

DEBUG:  distributed statement: INSERT INTO public.agg_events_102421 AS citus_table_alias (user_id, value_1_agg, value_4_agg) SELECT raw_events_first.value_4, raw_events_first.value_1, raw_events_first.user_id FROM public.raw_events_first_102413 raw_events_first WHERE (raw_events_first.user_id = 8) UNION SELECT raw_events_second.value_4, raw_events_second.value_1, raw_events_second.user_id FROM public.raw_events_second_102417 raw_events_second WHERE (raw_events_second.user_id = 8)

DEBUG:  distributed statement: INSERT INTO public.agg_events_102422 AS citus_table_alias (user_id, value_1_agg, value_4_agg) SELECT raw_events_first.value_4, raw_events_first.value_1, raw_events_first.user_id FROM public.raw_events_first_102413 raw_events_first WHERE (raw_events_first.user_id = 8) UNION SELECT raw_events_second.value_4, raw_events_second.value_1, raw_events_second.user_id FROM (SELECT NULL::integer AS user_id, NULL::timestamp without time zone AS "time", NULL::integer AS value_1, NULL::integer AS value_2, NULL::double precision AS value_3, NULL::bigint AS value_4 WHERE false) raw_events_second(user_id, "time", value_1, value_2, value_3, value_4) WHERE (raw_events_second.user_id = 8)


DEBUG:  distributed statement: INSERT INTO public.agg_events_102423 AS citus_table_alias (user_id, value_1_agg, value_4_agg) SELECT raw_events_first.value_4, raw_events_first.value_1, raw_events_first.user_id FROM public.raw_events_first_102413 raw_events_first WHERE (raw_events_first.user_id = 8) UNION SELECT raw_events_second.value_4, raw_events_second.value_1, raw_events_second.user_id FROM (SELECT NULL::integer AS user_id, NULL::timestamp without time zone AS "time", NULL::integer AS value_1, NULL::integer AS value_2, NULL::double precision AS value_3, NULL::bigint AS value_4 WHERE false) raw_events_second(user_id, "time", value_1, value_2, value_3, value_4) WHERE (raw_events_second.user_id = 8)

DEBUG:  distributed statement: INSERT INTO public.agg_events_102424 AS citus_table_alias (user_id, value_1_agg, value_4_agg) SELECT raw_events_first.value_4, raw_events_first.value_1, raw_events_first.user_id FROM public.raw_events_first_102413 raw_events_first WHERE (raw_events_first.user_id = 8) UNION SELECT raw_events_second.value_4, raw_events_second.value_1, raw_events_second.user_id FROM (SELECT NULL::integer AS user_id, NULL::timestamp without time zone AS "time", NULL::integer AS value_1, NULL::integer AS value_2, NULL::double precision AS value_3, NULL::bigint AS value_4 WHERE false) raw_events_second(user_id, "time", value_1, value_2, value_3, value_4) WHERE (raw_events_second.user_id = 8)
```

And, in the end, we insert 4 rows instead of 1 (if we have a single row with id = 8). I try to understand why this is happening. Do you have any ideas?